### ISSUE TITLE
Open-Source repos strategy

### ISSUE THREAD

**Issue by adrianthedev** [avo-hq/avo#2972]
# Proposed OSS solution

We need a solution for open source repositories like https://github.com/rubygems/rubygems.org

### Glossary

These are the stakeholders involved in this operation:

1. Repository admin; they have the gem server token and license key in their environment.
2. Anonymous developer; needs to clone, install, run, and run tests on the repo
3. CI. The Ci that runs the tests after each commit
3. The "Avo part of the app". This refers to the Avo engine, it's routes, business logic and views.
4. The rest of the app. This refers to everything that is not Avo (the front office) and it's business logic.
1. The parent app; This refers to the Rails app that's hosting the engine. It is comprised of the front-office (`main_app` and other engines) and the `avo` engine.

## Requirements

There are a few things we want to achieve with this proposal to maintain the ergonomics of open-source.

1. The Anonymous developer can clone the repo run `bundle install`, spin up the app and run the tests without a gem server token
1. The Anonymous developer does not have to have the gem server token installed
1. The Anonymous developer will be able to use the non-Avo parts of the app
1. The Anonymous developer will most likely not be able to use Avo
1. The Anonymous developer will not be able to run the Avo tests
2. The Repository admin will be able to see all parts (avo and non-Avo)
2. The Repository admin will be able to run all tests
2. The Repository admin will be able to work on the Avo side
1. The CI will have the token set in the environment variables
1. The CI will be able to run all the tests
1. The repository admin should not share the gem server token outside the organization so it won't leak the paid gems

## Proposal

1. Running `bundle install`

Our proposal here is to change a few things in our gem server and in the Avo internal in such a way that if some anonymous developer runs `bundle install`, they get dummy gems from our private gem server instead of the regular paid ones. This way we don't need to do any kind of `if/else` in the `Gemfile` or any other heueristic like that.

2. Running the app

THe app should boot up as expected with all the gems (`avo` and the dummy paid gems like `avo-advanced` and the rest). All users should be able to visit the front-office of the site.
The Avo side will most likely not work, or not work fully for an anonymous developer. That depends on the number of used APIs from the paid gems. If for example, there's a resource that is using the `scopes` method, thos won't work because that's an API introduced by a paid gem, but it will work perfectly for an repository admin.

3. Accessing the front-office (non-Avo)

All users will be able to visit the front-office of the app as usual as it doesn't interfere with the Avo engine.

4. Running tests

This is the bit that we need to apply some changes to the parent app. We need to conditionally run those tests based on a condition set on the environment.
We need to find a way to mark the Avo tests and run those only if the condition returns true.
After this conditional anyone can run the tests as usual, but only those admins which have the gem server token, and in turn, the paid gems installed, will have Avo tests ran.

5. Running in the CI

The CI will have it's own license key and gems server token, so it will be able to run all tests (Avo and non-Avo).

## Conclusion

We feel that applying this strategy we'll be able to keep the easy-to-run-and-install strategy of open source and keep the private parts of Avo away from being run by any developer with minimal invasive action on the parent app.

Looking forwards for feedback and things we might have missed.

Thanks,
Adrian ✌️

**Comment by simi** [comment]
> The Anonymous developer will most likely not be able to use Avo

Not ideal, but the way Avo 2 works is still OK.

> The Anonymous developer will not be able to run the Avo tests

This is IMHO no-go for projects like RubyGems.org. The only way to keep this acceptable would be to move Avo implementation outside of main repo and make it optional part or deployed on side. Any developer must be able to clone the project and run full test suite, otherwise there is risk outside contributions are not complete and useful.

**Comment by adrianthedev** [comment]
> Not ideal, but the way Avo 2 works is still OK.

Yes, unfortunately we can't expose the private gems in public repositories. That will give bad actors access to the code.

I understand your second point as well. We are open to suggestions on how to better support open source and still have the paid features available.

**Comment by nhorton** [comment]
@adrianthedev - I love Avo and just started using it on something when I realized I will have an issue with needing the Pro version, and the project is a Commercial Open Source one, thus hitting this issue.

My quick thoughts here are that not having the Pro gem available to everyone are that if:
1. If the code *runs* without the pro gem (ex. the "Menu customization is a pro feature" warning that happens now) and it has a setting that can be flipped to show the warning messages or just fully hide the disable functionality (i.e. what one might do in Dev and Test)
2. If there is a really easy test helper for rspec and minitest that gates tests on the Gem being there (i.e. `skip_unless_avo_pro` or such)
3. You have a good template Markdown file that can be dropped in a repo explaining what the Pro

Then I think the current situation might be fine. It means OSS devs can't have every test run locally, but when they put up a PR, it still can. This is similar to what happens with platforms in many cases - a dev can't always test Windows and Linux variants locally.

It also means that end-users of the software get a reasonable choice - pay a pretty modest fee to get the extra functionality or don't have some limits.

**Comment by adrianthedev** [comment]
Hey @nhorton.
Thanks for writing.

Are you saying that you can't run the Community tests locally without the Pro gem?
If that's correct, then it's a bug and we can/should fix it.
Or, maybe I misunderstood the message?

Before Avo 3 we had all the code in the main repo. The paid and free bits but we had a lot of bad actors cloning the repo and toggling off the licensing parts. We also had quite a bit of problems with people stealing the code outright. Someone who cloned it and planned to release it as MIT (the free and paid stuff).

Since then, we took the decision to protect the IP.
It's tough enough to start and maintain a business and having to fight those kinds of problems and that kind of "competition from our own code" is even tougher.

That's why we close-sourced the code.
From a maintainability perspective it's a PITA, but it is what it is.

Did that answer your question?
